src/compiler/checker.rs

```diff
@@ -457,13 +548,27 @@ impl TypeChecker {
         type_
     }
 
+    fn get_array_element_type_node(&self, node: &ArrayTypeNode) -> Option<Rc<Node /*TypeNode*/>> {
+        Some(node.element_type.clone())
+    }
+
     fn get_type_from_type_node(&self, node: &Node /*TypeNode*/) -> Rc<Type> {
         self.get_conditional_flow_type_of_type(self.get_type_from_type_node_worker(node), node)
     }
 
     fn get_type_from_type_node_worker(&self, node: &Node /*TypeNode*/) -> Rc<Type> {
-        match node.kind() {
-            SyntaxKind::NumberKeyword => self.number_type(),
+        let node = match node {
+            Node::TypeNode(type_node) => type_node,
+            _ => panic!("Expected TypeNode"),
+        };
```

Comment:
I bet you could make a `match_as!` macro to de-boilerplate this pattern a bit. (Probably a declarative macro would be fine for that?)

Reply:
Ya that would be a nice shorthand